### PR TITLE
fix(controller): dag and steps with unresolved with.params field. Fixes #7094

### LINKED ISSUE #7094: workflow fail if loop param results are skipped even if not used
## Summary

when you create a workflow with "withParam" provided by a previous steps, if you skip both steps, the workflow fail instantly with this error : 

```
withParam value could not be parsed as a JSON list: {{steps.generate.outputs.result}}: invalid character '{' looking for beginning of object key string
```

Argo workflow version 3.2.2, 3.3.3 

## Diagnostics
execute this workflow with `get-list` set to  "false" 

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: loops-param-result-
spec:
  arguments:
    parameters:
      - name: get-list
        value: true
  entrypoint: loop-param-result-example
  templates:
  - name: loop-param-result-example
    steps:
    - - name: generate
        when: "{{ workflow.parameters.get-list}} == true"
        template: gen-number-list
    - - name: sleep
        when: "{{ workflow.parameters.get-list}} == true" 
        
        template: sleep-n-sec
        arguments:
          parameters:
          - name: seconds
            value: "{{item}}"
        withParam: "{{steps.generate.outputs.result}}"
    - - name: sleep-2
        template: sleep-n-sec
        arguments:
          parameters:
          - name: seconds
            value: "5"

  - name: gen-number-list
    script:
      image: python:alpine3.6
      command: [python]
      source: |
        import json
        import sys
        json.dump([i for i in range(20, 31)], sys.stdout)
  - name: sleep-n-sec
    inputs:
      parameters:
      - name: seconds
    container:
      image: alpine:latest
      command: [sh, -c]
      args: ["echo sleeping for {{inputs.parameters.seconds}} seconds; sleep {{inputs.parameters.seconds}}; echo done"]

```

What Kubernetes provider are you using?
1.18
What executor are you running? Docker/K8SAPI/Kubelet/PNS/Emissary
k8s eks on aws

---
**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

### PR DESCRIPTION
**Pull request by ziv-codefresh**
Signed-off-by: ziv-codefresh <ziv@codefresh.io>

Fixes #7094

I found the same issue in dag template so this pr fixes it as well. 
ive added a test for each template (dag, steps) that should skip both of the steps and finish successfully. (previously the workflow was failed in the phase of resolving the with.params field)

important to mention - 
1. `expandStep/expandTask` running before `shouldExecute` function. means that regardless if the step will eventually run we are trying to parse `withParam` field
2.  with the following change if there is a step with unresolved `with.params` field the step will **not failed** as before, the step will be **skipped due to empty params condition**. this is kind of a change in the behaviour we have today, im not sure what is the correct approach here.  in case you trying to run some step with unresolved `withParam` - e.g `{{tasks.generate.outputs.result}}`the step will skipped rather than failed. let me know what do you think. thanks.

### CODE REVIEW

**Comment by ziv-codefresh** [comment]
@alexec @sarabala1979 PTAL. thanks!

**Comment by alexec** [comment]
Unresolved should be error. Empty should be skip?

**Comment by ziv-codefresh** [comment]
@alexec ive updated the description with additional information. 
`expandStep/expandTask` running before `shouldExecute` function. means that regardless if the step will eventually run we are trying to parse withParam field. 
so at this point we dont know if the step will run or not.. 
the issue is the following:
``when you create a workflow with "withParam" provided by a previous steps, if you skip both steps, the workflow fail instantly with this error``

with the change i suggested the step will be skipped rather than failed. 

the only solution i can think of is somehow is to check the parameters after we verify the step actually need to run. but im not sure what changes it will require. WDYT?

**Review by alexec — COMMENTED** [review]
> expandStep/expandTask running before shouldExecute function. 

Can we swap them around?

**Comment by ziv-codefresh** [comment]
i guess it is possible because all the steps/tasks that will be created (after the expending according to the `withParam` `WithItems` and `WithSequence`) will have the same when condition (correct me if im wrong).
so potentially we can check the when condition before we are expending them and only if its fulfilled we will continue to the actual expending, it will require some refactor in the `executeStepGroup` and `executeDAGTask`

File: workflow/controller/steps.go [hunk 1]
```diff
@@ -415,7 +394,29 @@ func (woc *wfOperationCtx) resolveReferences(stepGroup []wfv1.WorkflowStep, scop
 // expandStepGroup looks at each step in a collection of parallel steps, and expands all steps using withItems/withParam
 func (woc *wfOperationCtx) expandStepGroup(sgNodeName string, stepGroup []wfv1.WorkflowStep, stepsCtx *stepsContext) ([]wfv1.WorkflowStep, error) {
 	newStepGroup := make([]wfv1.WorkflowStep, 0)
+	stepTemplateScope := stepsCtx.tmplCtx.GetTemplateScope()
 	for _, step := range stepGroup {
+		childNodeName := fmt.Sprintf("%s.%s", sgNodeName, step.Name)
```
Comment (simster7) [inline workflow/controller/steps.go:399]: This should not be generated independently in two spots. If anything, this should be extracted into a helper function.

File: workflow/controller/steps.go [hunk 2]
```diff
@@ -415,7 +394,29 @@ func (woc *wfOperationCtx) resolveReferences(stepGroup []wfv1.WorkflowStep, scop
 // expandStepGroup looks at each step in a collection of parallel steps, and expands all steps using withItems/withParam
 func (woc *wfOperationCtx) expandStepGroup(sgNodeName string, stepGroup []wfv1.WorkflowStep, stepsCtx *stepsContext) ([]wfv1.WorkflowStep, error) {
 	newStepGroup := make([]wfv1.WorkflowStep, 0)
+	stepTemplateScope := stepsCtx.tmplCtx.GetTemplateScope()
```
Comment (simster7) [inline workflow/controller/steps.go:397]: Why lose the comment on this line?

File: workflow/controller/dag.go [hunk 3]
```diff
@@ -434,81 +434,82 @@ func (woc *wfOperationCtx) executeDAGTask(ctx context.Context, dagCtx *dagContex
...
-	expandedTasks, err := expandTask(*newTask)
-	if err != nil {
-		woc.initializeNode(nodeName, wfv1.NodeTypeSkipped, dagTemplateScope, task, dagCtx.boundaryID, wfv1.NodeError, err.Error())
-		connectDependencies(nodeName)
-		return
-	}
+	taskNodeName := dagCtx.taskNodeName(newTask.Name)
+	node = dagCtx.getTaskNode(newTask.Name)
+	var expandedTasks []wfv1.DAGTask
+	if node == nil {
```
Comment (simster7) [inline workflow/controller/dag.go:440]: I'm not comfortable with the refactor in this file. I don't think we should have so much (critical) logic controlled by nested if else statements. Is there a way to refactor it such that the `shouldExecute` check happens outside in its own block? Is it possible to have it as a normal block before the `expandedTasks, err := expandTask(*newTask)` line as it currently exists on the master branch?

**Review by simster7 — CHANGES_REQUESTED** [review]
In my opinion this change, even though it's a simple reordering, has a large surface exposure and should be carefully implemented and reviewed

**Comment by sarabala1979** [comment]
@ziv-codefresh it should fix in `ProcessArgs` if any variable is not resolved or substituted, it should take the default value.
If the default value is not configured, the workflow should fail with `not supplied`

**Comment by ziv-codefresh** [comment]
@sarabala1979 @simster7 sorry guys its not ready to review.
i had conversion above it @alexec and i suggest some solution and opened the pr but seems its not the correct approach.

its just some WIP that i tried to refactor it but its far from be ready. ill mark this as a draft. sorry for that.

**Comment by sarabala1979** [comment]
@ziv-codefresh no issues. I think this should be fixed globally.  Currenly `withParam` is failling because of variable is not substituted others will just print the actual variable string `{{xxx}}`.  `SubstituteParams` is allowing the unresolved variables `, err := template.Replace(globalReplacedTmplStr, replaceMap, true)`. 

Suggestion:
  if any variable is not resolved or substituted, it should take the default value.
If the default value is not configured, the workflow should fail with not supplied.

**Comment by ziv-codefresh** [comment]
@sarabala1979 sounds good. so you are suggesting that every substation will use the `ProcessArgs` method? 
btw, i dont think it will solve the issue which raised here, the issue is the following:
`when you create a workflow with "withParam" provided by a previous steps, if you skip both steps, the workflow fail instantly`
so basically if im expecting that step X will run and then step Y and step Y use the output of X , so even if both of the steps should be skipped, it will failed in the phase of expanding (in case the user will did not set any default value) 
am i missing something?

**Comment by sarabala1979** [comment]
@ziv-codefresh you are trying to add logic if the current step is referring previous step variable and the previous step is skipped/didn't run.  workflow should fail. Am I getting correctly?

**Comment by ziv-codefresh** [comment]
@sarabala1979 actually no, if the current step is running and referring previous step variable and the previous step is skipped/didn't run the workflow will fail (as expected).
the issue is when the current step should not run (because his when condition is not fulfilled) and this step referring previous step variable and the previous step is skipped/didn't run the workflow will fail and this is wrong behaviour.
the reason that the workflow will fail in this case is because we are trying to expand the task before we are checking if the step should run. so basically what happens is that we are failing the all workflow because we are enable to resolve the withParams field for a step that will not run at all.
to solve this, we need first of all to check if the step should run, and only if he should run we need to apply the expand logic.
by changing this we will get the following:
if B is running and referring to step A variable and step A is skipped/didn't run the workflow will fail (As today).
if B should should be skipped and referring to step A variable and step A is skipped/didn't run the workflow will not fail and step B will be skipped.
```
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: loops-param-result-
spec:
  arguments:
    parameters:
      - name: get-list
        value: false
  entrypoint: loop-param-result-example
  templates:
  - name: loop-param-result-example
    steps:
    - - name: generate
        when: "{{ workflow.parameters.get-list}} == true"
        template: gen-number-list
    - - name: sleep
        when: "{{ workflow.parameters.get-list}} == true" 
        
        template: sleep-n-sec
        arguments:
          parameters:
          - name: seconds
            value: "{{item}}"
        withParam: "{{steps.generate.outputs.result}}"
```
today, this workflow will fail because `steps.generate.outputs.result` will not be resolved (generate step will be skipped)
but also `sleep` should be skipped and there is no reason the workflow will failed.
swapping the logic of `shouldExecute` and `expandTask` should solve it, because we first of all check if the step should run, and only if it is we run the expand logic. 
i hope it more clear now.

**Comment by ziv-codefresh** [comment]
@alexec @sarabala1979 @simster7 so i tried to swap the `shouldExecute` and `expandTask/expandDag` functions, but i dont see how it can work. 
`shouldExecute` function needs a substituted when expression to decide whether or not a step should execute, and the `expandTask/expandDag` needs to know if the step actually need to run in order to perform the substation.
so its kind of `chicken egg` problem here and im not sure what can be done. 
example for problematic template 
```
....
    - - name: divide-by-2
        template: divide-by-2
        arguments:
          parameters:
          - name: num
            value: "{{item.num}}"
        withParam: "{{steps.odd-or-even.outputs.parameters}}"
        when: "{{item.evenness}} == even"
...

```
so in order to calculate the when condition we need to resolve `{{steps.odd-or-even.outputs.parameters}}` field.
but in order to decide if we need to resolve `{{steps.odd-or-even.outputs.parameters}}` we need to calculate then when condition.

**Comment by ziv-codefresh** [comment]
@alexec @sarabala1979 i would appreciate if you can take a quick look here.

> @alexec @sarabala1979 @simster7 so i tried to swap the `shouldExecute` and `expandTask/expandDag` functions, but i dont see how it can work. `shouldExecute` function needs a substituted when expression to decide whether or not a step should execute, and the `expandTask/expandDag` needs to know if the step actually need to run in order to perform the substation. so its kind of `chicken egg` problem here and im not sure what can be done. example for problematic template
> 
> ```
> ....
>     - - name: divide-by-2
>         template: divide-by-2
>         arguments:
>           parameters:
>           - name: num
>             value: "{{item.num}}"
>         withParam: "{{steps.odd-or-even.outputs.parameters}}"
>         when: "{{item.evenness}} == even"
> ...
> ```
> 
> so in order to calculate the when condition we need to resolve `{{steps.odd-or-even.outputs.parameters}}` field. but in order to decide whether or not to resolve `{{steps.odd-or-even.outputs.parameters}}` we need to calculate the `when` condition.

**Pull request closed without merging**